art/estimators/object_detection/pytorch_yolo.py

```diff
@@ -0,0 +1,490 @@
+
+        elif isinstance(x, np.ndarray):
+            x_preprocessed, y_preprocessed = self._apply_preprocessing(x, y=y, fit=False, no_grad=True)
+
+            if y_preprocessed is not None and isinstance(y_preprocessed[0]["boxes"], np.ndarray):
+                y_preprocessed_tensor = []
+                for i, y_i in enumerate(y_preprocessed):
+                    y_preprocessed_t = {
+                        "boxes": torch.from_numpy(y_i["boxes"]).type(torch.float).to(self.device),
+                        "labels": torch.from_numpy(y_i["labels"]).type(torch.int64).to(self.device),
+                    }
+                    if "masks" in y_i:
+                        y_preprocessed_t["masks"] = torch.from_numpy(y_i["masks"]).type(torch.uint8).to(self.device)
+                    y_preprocessed_tensor.append(y_preprocessed_t)
+                y_preprocessed = y_preprocessed_tensor
+
+            if self.clip_values is not None:
+                norm_factor = self.clip_values[1]
+            else:
+                norm_factor = 1.0
+
+            x_grad = torch.from_numpy(x_preprocessed / norm_factor).to(self.device)
+            x_grad.requires_grad = True
+            image_tensor_list_grad = x_grad
+            inputs_t = image_tensor_list_grad
+
+        else:
+            raise NotImplementedError("Combination of inputs and preprocessing not supported.")
+
+        labels_t = translate_labels_art_to_yolov3(labels_art=y_preprocessed)
+
+        loss_components = self._model(inputs_t, labels_t)
+
+        return loss_components, inputs_t, image_tensor_list_grad
+
+    def loss_gradient(  # pylint: disable=W0613
+        self, x: np.ndarray, y: List[Dict[str, Union[np.ndarray, "torch.Tensor"]]], **kwargs
+    ) -> np.ndarray:
+        """
+        Compute the gradient of the loss function w.r.t. `x`.
+
+        :param x: Samples of shape (nb_samples, height, width, nb_channels).
+        :param y: Target values of format `List[Dict[Tensor]]`, one for each input image. The
+                  fields of the Dict are as follows:
+
+                  - boxes (FloatTensor[N, 4]): the predicted boxes in [x1, y1, x2, y2] format, with values \
+                    between 0 and H and 0 and W
+                  - labels (Int64Tensor[N]): the predicted labels for each image
+                  - scores (Tensor[N]): the scores or each prediction.
+        :return: Loss gradients of the same shape as `x`.
+        """
+        output, inputs_t, image_tensor_list_grad = self._get_losses(x=x, y=y)
+
+        # Compute the gradient and return
+        loss = None
+        for loss_name in self.attack_losses:
+            if loss is None:
+                loss = output[loss_name]
+            else:
+                loss = loss + output[loss_name]
+
+        # Clean gradients
+        self._model.zero_grad()
+
+        # Compute gradients
+        loss.backward(retain_graph=True)  # type: ignore
+
+        if isinstance(x, np.ndarray):
```

Comment:
from the type hinting x can only be a np.ndarray. Is this if statement redundant or can x be other types (tensors, etc)?

Reply:
The typing for `x` should include tensors, I have updated the typing.